Modularise MTU discovery

bimmlerd opened the pull request:
This series is a preparation for loader modularisation, by pulling a little bit of IPsec key setup and MTU into hive, so that the future loader cell can consume it.

bimmlerd committed:
ipsec/daemon: pull ipsec setup from dmn startup

Introduces the IPsec key custodian which encompasses the small part of
IPsec initialization logic which was in daemon, with the goal of
preparing MTU modularization. MTU depends on knowing the IPsec key
size (though, notably, it seemingly assumes the size doesn't change),
hence we start here with the modularisation.

This doesn't modularise a significant part of the IPsec subsystem, but
it's a incremental step in the direction. It would be conceivable to
also pull the key file watcher into the custodian, but I've opted for
the smallest diff here.

In terms of initialization ordering, this puts the IPsec key setup
_before_ the daemon startup, but after the local node store. As a
result, the SPI is guaranteed to be set earlier than before this patch.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
mtu: create MTU discovery cell

Create a minimal MTU discovery cell which exposes an MTU interface
capable of fulfilling the MTU queries present in current-day code. This
is a pre-requisite to modularise the loader, and itself depends on the
initial modularisation of the IPsec key setup steps.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

brb reviewed (approved):
:+1: WireGuard changes.

pchaigno commented on daemon/cmd/daemon.go:
```diff
@@ -424,11 +424,6 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
 	}
```
> MTU depends on knowing the IPsec key size (though, notably, it seemingly assumes the size doesn't change)

That's a bug. Since https://github.com/cilium/cilium/pull/19814, we allow users to change the IPsec secret without restarting the agent. If they change the authentication key size at that time, then we may end up with an incorrect MTU.

At the very least, we'd need an error message and a mention in the documentation. Do you want to take care of that or should I ask for someone to be assigned?

bimmlerd replied:
It is. I don't have enough context to deal with this myself, though we are laying the ground-work for this kind of reactivity in the agent with the runtime device work at the moment. If you could ask for someone, that'd be appreciated.

pchaigno replied:
Will do. Thanks for reporting it!

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -1012,7 +967,54 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
+				return nil
+			}
+		}
+	}))
```
Could we split this second commit into several to ease conflict resolution during rebase and backporting of other IPsec work?

In particular, it currently seems to mix code movements, function prototype changes/renaming, health reporting, and misc. other changes. I'd like to see the large code movement eliminated or moved to its own commit with only that. It's what is most likely to conflict and will be much easier to confidently resolve if we can just reapply the same movement without needing to watch out for 2 other types of changes.

bimmlerd replied:
I've split the second commit in a few smaller commits, PTAL.

pchaigno reviewed (changes requested):
Thanks for the ping David!

The second commit seems the most likely to conflict with other ongoing IPsec work. Could we split it into smaller, easier-to-handle commits? Or alternatively, skip it if that seems easier (IIUC, it's not needed by the third commit)?

bimmlerd committed:
ipsec: move starting jobs into IPsecKeyCustodian

With the introduction of this new IPsecKeyCustodian cell, it makes sense
to move the responsibility of starting the background workers to the
cell's interface. This move additionally prepares the code for
conversion to the hive/job framework.

Ideally, this would not still live in cmd/daemon, but in the start hook
of the key custodian, but I'm unsure of the ordering dependencies that
exist, and don't want to rock the boat too badly at this stage.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
ipsec: wrap NodeDiscovery dependency in interface

In preparation of changing the key file watcher functionality into hive jobs,
weaken the explicit dependency on NodeDiscovery and replace it with
the (newly introduced) NodeUpdater interface, as this is the only functionality
the KeyfileWatcher needs from NodeDiscovery.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
ipsec: convert goroutines to hive jobs

Instead of launching raw goroutines, we can employ the hive job
mechanism. As is, this doesn't give an immediate benefit (except for
aligning with project direction), but enables health reporting in a
follow-up commit.

The (minor) functional difference is that instead of waiting for a
minute in between the stale key reclamation, we now start the job every
minute, as long as it completes within that minute.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
ipsec: drop now-unused daemon context

Previous refactoring removed the dependency on the daemon context, hence
we can drop it from the interface and implementations.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
ipsec: add health reporting to keyfile watcher

Use the job health reporting API to add visibility to the health of the
keyfile watcher.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

pchaigno reviewed (approved):
Perfect! Thanks a lot David :pray:
